port/cpl_float.cpp

```diff
@@ -306,3 +309,129 @@ float CPLConvertHalfToFloat(GUInt16 nHalf)
     std::memcpy(&fFloat32, &nFloat32, sizeof fFloat32);
     return fFloat32;
 }
+
+namespace
+{
+
+template <typename T> struct Fraction
+{
+    using value_type = T;
+
+    T num;
+    T denom;
+};
+
+/** Approximate a floating point number as a fraction, using the method describe
+ * in Richards, Ian (1981). Continued Fractions Without Tears. Mathematics
+ * Magazine, Vol. 54, No. 4. https://doi.org/10.2307/2689627
+ *
+ * If the fraction cannot be approximated within the specified error tolerance
+ * in a certain amount of iterations, a warning will be raised and  std::nullopt
+ * will be returned.
+ *
+ * @param x the number to approximate as a fraction
+ * @param err the maximum allowable absolute error in the approximation
+ *
+ * @return the approximated value, or std::nullopt
+ *
+*/
+std::optional<Fraction<std::uint64_t>> FloatToFraction(double x, double err)
+{
+    using inttype = std::uint64_t;
+    int nMaxIter = 1000000;
```

Comment:
That's a lot of iterations. Do we really need that much ? Although I suspect 1 million for modern CPU is a fraction of second...